### DISCUSSION TITLE
TextArea default height

### DISCUSSION THREAD

**Discussion by HartiganHM** [twilio-labs/paste#1569] [Q&A] [answered]
Hey, all!

We're running the latests `@twilio-paste/core` which is `9.6.1` and have a couple of question about the default `height/minHeight` of the `TextArea` component. In the docs, it looks like this is set to `76px`:

![image](https://user-images.githubusercontent.com/24458700/122112670-50fb1c80-cdde-11eb-95f1-07aad9be2f9b.png)

However, when we go to implement it, it defaults to `36px`:

![image](https://user-images.githubusercontent.com/24458700/122112719-5fe1cf00-cdde-11eb-8ce4-e3033be86004.png)

I've double-checked any cascading styles, tried it outside of our current page, and copy-pasted code snippets directly from the docs and it stays at `36px`.

Is this expected? If so, how do we visually distinguish a "text area" form field from an "input" form field, especially when the `height: never` is set on the `TextAreaProps`?

Happy to share more details as needed, thanks!

**Comment by richbachman** [comment, marked as answer]
Hi @HartiganHM. That looks like a tricky one. Sorry for the troubles. I just setup a CodeSandbox using TextArea and it looks like its showing the height as expected: https://codesandbox.io/s/solitary-microservice-7tv7l?file=/src/index.tsx.

Can you share any other information about your implementation so we can help troubleshoot. Thanks!

**Reply by HartiganHM** [reply]
Thanks for the quick response, @richbachman!

This is more rubber ducking than anything, but maybe you can spot something I can't. A few things I've tried:

- We're in a legacy code base, so I tried looking for something that might be conflicting in a stylesheet, but didn't find anything
- We're using the "sendgrid" theme with `Theme.Provider`, but I updated the Sandbox and there was no change
- I noticed the styles are being applied in a `style` tag, but couldn't find it anywhere in our repo
- I noticed Paste is using `react-textarea-autosize@7.1.0` and the component _does_ update its `height` with each new row, so this seems to be working as expected
  - _Note: It does seem that this is how `react-textarea-autosize` updates the height, so maybe this is the culprit?_
- I checked the all "height" style props in our app versus Sandbox and the only difference is the `element.style.height` prop (ours, left; Sandbox, right)

> ![image](https://user-images.githubusercontent.com/24458700/122120149-46915080-cde7-11eb-94a4-86f5b4bf1b9f.png)

**Reply by HartiganHM** [reply]
Hey, @richbachman. So, after more investigation, I've been able to determine that this is possibly due (at least, in part) to `react-textarea-autosize`. By following [these steps](https://stackoverflow.com/a/29658014) I was able to see when the `element.style` attribute of the `TextArea` was updated and the below was included in the file where the Chrome Dev Tools stopped in the "Sources" tab:

```
/*!
  autosize 4.0.2
  license: MIT
  http://www.jacklmoore.com/autosize
*/
```

Looking at the Changelog for the `TextArea` component, I thought I could try a different version, so I added `@twilio-paste/textarea@3.0.2` since that seems like [the version](https://paste.twilio.design/components/textarea/#3-0-2) where compatibility for `react-hook-form` was included, which we're using, and before some dependency updates.

This, too, did not work 😅 

Finally I just started deleting things in the legacy app and it was fixed, so now I'm headed down the rabbit hole 🐰 🕳️ 

I think we can safely close this ticket. Thanks for the help!

**Reply by richbachman** [reply]
Hey @HartiganHM. Thanks so much for going down the rabbit hole here. So from what I'm gathering its a conflict with the legacy codebase and `react-textarea-autosize`, right?

If that's it, let us know what you find out in case we run into this again.

**Reply by HartiganHM** [reply]
So, the codebase we're working with is pulling styles from all over the place. It was hard to narrow down, but something was telling `react-textarea-autosize` that the "default" or "auto" height for a `textarea` element was `36px`. Because of this `react-textarea-autosize` was setting this as the initial height. 

In other words, the root issue was with the legacy code base, not `react-textarea-autosize`.

The fix was to add the below code underneath all of the legacy style imports so it was applied last in the cascade. This effectively reset the `height` to whatever the Paste component was expecting:

```css
textarea {
  height: auto;
}
```

**Reply by richbachman** [reply]
Good to know. Thanks for adding the solution. Out of curiosity what codebase did you have to add this fix to?

**Reply by HartiganHM** [reply]
We're rewriting the [SendGrid Help Center](https://support.sendgrid.com/hc/en-us). The legacy codebase is a mix of Ruby/CoffeScript/SCSS and we're updating to use React/CSS Modules. 

We need to inject the new app into the legacy app, however, so there have been some odd styling wires crossed, especially because the legacy app uses a custom-built compiler (versus webpack, etc.).

I can DM you the internal codebase if you're curious on looking at the code, so just LMK!

**Reply by richbachman** [reply]
Yeah. That would be great. Thanks!

**Reply by SiTaggart** [reply]
@HartiganHM Curious what the plan or need is for CSS Modules? Are you going all in on Paste?

**Reply by HartiganHM** [reply]
@richbachman Sent!

Hey, @SiTaggart! CSS Modules is a team adopted practice for styling. I've been with the team since August and it was cemented before I joined. I don't think we have any plans to change, but I do know that we're going as all in on Paste as is possible. 

Would CSS Modules provide any issues? Do you have another recommendation?

**Reply by SiTaggart** [reply]
It wouldn't necessarily cause any direct conflicts with Paste, but you might run into situations where you have a less than ideal customer experience if you build some custom UI using CSS modules, but also want to take advantage of something like a dark mode.

The CSS modules part of the UI wouldn't respond to the theme change in the same way as the other Paste components, leading to a sub-optimal customer  experience. We're hoping you get everything you need from Paste and don't need to use something like CSS modules, that way your custom UI will also respond to the same theme and design language updates, just like out of the box component would. 

We have produced a short guide on this https://paste.twilio.design/tokens/how-to-compose-custom-ui-with-tokens

**Reply by HartiganHM** [reply]
Yeah, that all makes sense. We do have some need for custom components outside of what Paste currently provides. For instance, I'm working on a `FileInput` right now that we need in the next few days. However, it's using a Paste `Button` and the native `input` is hidden, so we don't anticipate any cascading side-effects. 

Additionally, for some of our devs (myself included) it's easier to some styling (e.g. layouts) in CSS/SCSS. We're aware of things like `Flex/Grid` and have tried them out, but CSS gives us a bit more flexibility (though we have divergent opinions about this on our team, so I'm just speaking for myself).

The guide you provided should be helpful, so I'll make sure to reference it for any custom component we build. Happy to discuss things more offline if you're curious about our approach/have suggestions!
